platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -117,180 +118,198 @@ - (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshot
-            _loading = false;
+    __weak __typeof__(self) weakSelf = self;
+    _snapshotCallback = std::make_unique<mbgl::Actor<mbgl::MapSnapshotter::Callback>>(*mbgl::Scheduler::GetCurrent(), [=](std::exception_ptr mbglError, mbgl::PremultipliedImage image, mbgl::MapSnapshotter::Attributions attributions, mbgl::MapSnapshotter::PointForFn pointForFn) {
+        weakSelf.loading = false;
+        
+        
+        if (mbglError) {
+            NSString *description = @(mbgl::util::toString(mbglError).c_str());
+            NSDictionary *userInfo = @{NSLocalizedDescriptionKey: description};
+            NSError *error = [NSError errorWithDomain:MGLErrorDomain code:MGLErrorCodeSnapshotFailed userInfo:userInfo];
             
-            NSMutableArray *infos = [NSMutableArray array];
+            // Dispatch result to origin queue
+            dispatch_async(queue, ^{
+                _snapshotCallback = NULL;
+                completion(nil, error);
+            });
+        } else {
+            // This code is in place to test the _snapshotCallback release cycle.
+            dispatch_async(queue, ^{
+                _snapshotCallback = NULL;
+                completion(nil, nil);
+            });
 
 #if TARGET_OS_IPHONE
-            CGFloat fontSize = [UIFont smallSystemFontSize];
-            UIColor *attributeFontColor = [UIColor blackColor];
+//            MGLImage *mglImage = [[MGLImage alloc] initWithMGLPremultipliedImage:std::move(image) scale:weakSelf.options.scale];
 #else
-            CGFloat fontSize = [NSFont systemFontSizeForControlSize:NSMiniControlSize];
-            NSColor *attributeFontColor = [NSColor blackColor];
+            MGLImage *mglImage = [[MGLImage alloc] initWithMGLPremultipliedImage:std::move(image)];
+            mglImage.size = NSMakeSize(mglImage.size.width / weakSelf.options.scale,
+                                       mglImage.size.height / weakSelf.options.scale);
 #endif
-            for (auto attribution = attributions.begin(); attribution != attributions.end(); ++attribution) {
-                NSString *attributionHTMLString = @(attribution->c_str());
-                NSArray *tileSetInfos = [MGLAttributionInfo attributionInfosFromHTMLString:attributionHTMLString
-                                                                                  fontSize:fontSize
-                                                                                 linkColor:attributeFontColor];
-                [infos growArrayByAddingAttributionInfosFromArray:tileSetInfos];
-            }
-            
-            _attributionInfo = infos;
+//            [weakSelf drawAttributedSnapshot:attributions snapshotImage:mglImage pointForFn:pointForFn queue:queue completionHandler:completion];
```

Comment:
The code for snapshotter used to be part of this method. I move it to it's own method to isolate each memory leak that I detected.